Integration of FinoaConnect wallet connector service into Web3Onboard

finoaamukherjee opened the pull request:
### Description
Integration of FinoaConnect (Finoa's Wallet Connector Service) into Web3Onboard for distribution through BlockNative. For any questions or queries, please reach out to:
* Anirban Mukherjee (anirban.mukherjee@finoa.io)
* Laura Pescatori (laura.pescatori@finoa.io)


#### **_PLEASE NOTE- Checklist must be complete prior to review._**
### Checklist
- [X] Increment the version field in `package.json` of the package you have made changes in following [semantic versioning](https://semver.org/) and using alpha release tagging
- [X] Check the box that allows repo maintainers to update this PR
- [X] Test locally to make sure this feature/fix works
- [X] Run `yarn check-all` to confirm there are not any associated errors
- [ ] Confirm this PR passes Circle CI checks
- [X] Add or update relevant information in the documentation

### Docs Checklist
- [X] Include a screenshot of any changes ([see docs README on running locally](https://github.com/blocknative/web3-onboard/blob/develop/docs/README.md))
- [X] Add/update the appropriate package README (if applicable)
- [X] Add/update the related module in the [docs demo](https://github.com/blocknative/web3-onboard/blob/develop/docs/src/lib/services/onboard.js) (if applicable)
- [X] Add/update the related package in the `docs/package.json` file (if applicable)

### If this PR includes changes to add an injected wallet or SDK wallet module: 
Please complete the following using the internal demo package.
To run this demo use the command `yarn && yarn dev` to get the project running at `http://localhost:8080/`
[Finoa] It was verified that Finoa's documentation appears under `docs`. The `FinoaConnect` option does show up on the list of injected wallets, and we have verified that connecting and disconnecting the wallet to the dApp works fine.


#### Tests with demo app (injected)
- [ ] send transaction
- [ ] switch chains
- [ ] sign message
- [ ] sign typed message
- [X] disconnect

#### Tests with demo app (SDK)
- [x] send transaction
- [x] switch chains
- [x] sign message
- [x] sign typed message
- [X] disconnect

### Screenshots

Screenshot of FinoaConnect under `docs`
<img width="1552" alt="Screenshot 2024-05-06 at 14 48 06" src="https://github.com/blocknative/web3-onboard/assets/168822311/eaf50326-3b4c-4bfa-b189-4a101b9a438c">

finoaamukherjee commented:
There are 2 topics pending on this PR on which we would appreciate some guidance:
* (Circle) CI pipeline is breaking in a job to compile `core` package, but the changes we have made in this PR does not include any changes to this package.
* * Could you provide some guidance on why this might be happening ?
* We have tried to include "@web3-onboard/finoaconnect:1.0.0-alpha.1" as a package in the `docs` but that fails in yarn checks. We are not sure why this might be happening. 
* * Could you provide some guidance on how to fix this?

Adamj1232 commented on packages/finoaconnect/package.json:
```diff
@@ -0,0 +1,66 @@
+{
+    "name": "@web3-onboard/finoaconnect",
+    "version": "1.0.0-alpha.1",
```
```suggestion
    "version": "2.0.0-alpha.1",
```

Adamj1232 commented on packages/finoaconnect/src/index.ts:
```diff
@@ -0,0 +1,78 @@
+import { FinoaEIP1193Provider } from '@finoa/finoa-connect-sdk'
+import type { Chain, WalletInit } from '@web3-onboard/common'
+
+export interface FinoaWalletOption {
```
Adding this type to the docs and readme would be helpful for devs

Adamj1232 replied:
Also adding notes on what the props are would be helpful

finoaamukherjee replied:
done. Any thoughts?

Adamj1232 commented on packages/finoaconnect/src/index.ts:
```diff
@@ -0,0 +1,78 @@
+                        chain => BigInt(chain.id) === chainIdN
+                      )
+                    } catch {
+                      /* empty */
```
Would this error need to be handled?

finoaamukherjee replied:
no need to handle this error. It is a case when the parent error does not have a valid chain id.
Updated the comment to be more expressive

Adamj1232 commented on packages/demo/package.json:
```diff
@@ -36,6 +36,7 @@
     "@web3-onboard/coinbase": "^2.2.7",
     "@web3-onboard/dcent": "^2.2.7",
     "@web3-onboard/enkrypt": "^2.0.3",
+    "@web3-onboard/finoaconnect": "1.0.0-alpha.1",
```
```suggestion
    "@web3-onboard/finoaconnect": "2.0.0-alpha.1",
```

Adamj1232 commented:
> There are 2 topics pending on this PR on which we would appreciate some guidance:
> 
> * (Circle) CI pipeline is breaking in a job to compile `core` package, but the changes we have made in this PR does not include any changes to this package.
> * * Could you provide some guidance on why this might be happening ?
> * We have tried to include "@web3-onboard/finoaconnect:1.0.0-alpha.1" as a package in the `docs` but that fails in yarn checks. We are not sure why this might be happening.
> * * Could you provide some guidance on how to fix this?

Thanks for the PR - I have added a few comments and questions.

Ill checkout the CI pipeline issue.
As for testing the implementation you will use packages/demo, which will need the new module added to the package.json, initialized and added to the demo code. Then you can run `yarn && yarn dev` to get the demo started and complete the PR checklist.

Thanks again!

Adamj1232 commented on .circleci/config.yml:
```diff
@@ -732,6 +738,12 @@ jobs:
     working_directory: ~/web3-onboard-monorepo/packages/particle-network
     steps:
       - node-staging-build-steps
+  build-finoaconnect:
+    docker:
+      - image: cimg/node:18.0.0
+    working_directory: ~/web3-onboard-monorepo/packages/finoaconnect
+    steps:
+      - node-build-steps
```
```suggestion
  build-staging-finoaconnect:
    docker:
      - image: cimg/node:18.0.0
    working_directory: ~/web3-onboard-monorepo/packages/finoaconnect
    steps:
      - node-staging-build-steps
```

Adamj1232 commented on packages/finoaconnect/README.md:
```diff
@@ -0,0 +1,39 @@
+`npm i @web3-onboard/core @web3-onboard/finoaconnect`
+
+## Usage
+
```
```suggestion

Optional initialization object
```typescript
/** Optional object provided to the initiation of the wallet connector. 
 * When not included, the wallet connector service connects to FinoaConnect production systems.
 * @field {url} URL of the FinoaConnect backend systems to be used
 * @field {labelSuffix} arbitrary string label to denote the context of the URL field  */
export interface FinoaWalletOption {
  url?: string
  labelSuffix?: string
}
```

```

Adamj1232 commented on docs/src/routes/docs/[...4]wallets/[...36]finoaconnect/+page.md:
```diff
@@ -0,0 +1,45 @@
+</Tabs>
+
+## Usage
+
```
```suggestion

Optional initialization object
```typescript
/** Optional object provided to the initiation of the wallet connector. 
 * When not included, the wallet connector service connects to FinoaConnect production systems.
 * @field {url} URL of the FinoaConnect backend systems to be used
 * @field {labelSuffix} arbitrary string label to denote the context of the URL field  */
export interface FinoaWalletOption {
  url?: string
  labelSuffix?: string
}
```

```

Adamj1232 replied:
You may need to fix the triple tick to close that code block since github doesnt like it in the suggestion but thats the general idea to ensure users know about the init object

finoaamukherjee replied:
Sorry, I do not follow you here. The code segment (denoted with triple-ticks) is closed in line 45.
The  additional "typescript" word in the code section provides custom colouring on IDE's where recognition of typescript keywords are supported.

Adamj1232 replied:
I added the code changes

finoaamukherjee commented:
> > There are 2 topics pending on this PR on which we would appreciate some guidance:
> > 
> > * (Circle) CI pipeline is breaking in a job to compile `core` package, but the changes we have made in this PR does not include any changes to this package.
> > * * Could you provide some guidance on why this might be happening ?
> > * We have tried to include "@web3-onboard/finoaconnect:1.0.0-alpha.1" as a package in the `docs` but that fails in yarn checks. We are not sure why this might be happening.
> > * * Could you provide some guidance on how to fix this?
> 
> Thanks for the PR - I have added a few comments and questions.
> 
> Ill checkout the CI pipeline issue. As for testing the implementation you will use packages/demo, which will need the new module added to the package.json, initialized and added to the demo code. Then you can run `yarn && yarn dev` to get the demo started and complete the PR checklist.
> 
> Thanks again!

The checks for sanity are now completed in the PR, except the one for CI pipelines.
The dApp on localhost boots up fine, but does not enlist the FinoaConnect option yet. I will work to checkout why this is happening.

Adamj1232 commented on packages/finoaconnect/README.md:
````diff
@@ -0,0 +1,51 @@
+
+Optional initialization object
+```typescript
+	/** Optional object provided to the initiation of the wallet connector. 
+	 * When not included, the wallet connector service connects to FinoaConnect production systems.
+	 * @field {url} URL of the FinoaConnect backend systems to be used
+	 * @field {labelSuffix} arbitrary string label to denote the context of the URL field  */
+	export interface FinoaWalletOption {
+	  url?: string
+	  labelSuffix?: string
+	}
````
```suggestion
/** Optional object provided to the initiation of the wallet connector. 
 * When not included, the wallet connector service connects to FinoaConnect production systems.
 * @field {url} URL of the FinoaConnect backend systems to be used
 * @field {labelSuffix} arbitrary string label to denote the context of the URL field  */
export interface FinoaWalletOption {
  url?: string
  labelSuffix?: string
}
```

Adamj1232 commented on docs/src/routes/docs/[...4]wallets/[...36]finoaconnect/+page.md:
````diff
@@ -0,0 +1,57 @@
+
+Optional initialization object
+```typescript
+	/** Optional object provided to the initiation of the wallet connector. 
+	 * When not included, the wallet connector service connects to FinoaConnect production systems.
+	 * @field {url} URL of the FinoaConnect backend systems to be used
+	 * @field {labelSuffix} arbitrary string label to denote the context of the URL field  */
+	export interface FinoaWalletOption {
+	  url?: string
+	  labelSuffix?: string
+	}
````
```suggestion
/** Optional object provided to the initiation of the wallet connector. 
 * When not included, the wallet connector service connects to FinoaConnect production systems.
 * @field {url} URL of the FinoaConnect backend systems to be used
 * @field {labelSuffix} arbitrary string label to denote the context of the URL field  */
export interface FinoaWalletOption {
  url?: string
  labelSuffix?: string
}
```

Adamj1232 commented on packages/finoaconnect/package.json:
```diff
@@ -0,0 +1,66 @@
+      "type-check": "tsc --noEmit"
+    },
+    "devDependencies": {
+      "typescript": "5.1.6"
```
Can we update to the latest version of typescript here? We have a feature in the pipeline that will require this

finoaamukherjee replied:
resolved in latest commit

finoaamukherjee commented:
![Screenshot 2024-05-15 at 10 51 42](https://github.com/blocknative/web3-onboard/assets/168822311/0ea0f596-1337-45f8-b674-59f3cd81da4c)
![Screenshot 2024-05-15 at 11 44 28](https://github.com/blocknative/web3-onboard/assets/168822311/c84375ed-a8d6-4987-b0f3-f7132583eff8)
![Screenshot 2024-05-15 at 11 55 47](https://github.com/blocknative/web3-onboard/assets/168822311/83a05a97-5419-4166-a2c6-1eaea68792ab)
![Screenshot 2024-05-15 at 11 56 33](https://github.com/blocknative/web3-onboard/assets/168822311/bb23a089-8301-4fdf-aebf-42b8052d45a5)

finoaamukherjee commented:
The PR has now been updated to return error code 4200 (UNSUPPORTED_METHOD) for all method calls that are not supported by FinoaConnect provider.

Adamj1232 commented on packages/finoaconnect/src/index.ts:
```diff
@@ -0,0 +1,78 @@
+                ...args: Parameters<FinoaEIP1193Provider['request']>
+              ) {
+
+                // if (args[0].method === "eth_getBalance") {
```
Can we remove commented out code?

Adamj1232 commented on packages/finoaconnect/yarn.lock:
```diff
@@ -0,0 +1,3590 @@
+# THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.
```
This file can be removed all together

Adamj1232 commented:
@finoaamukherjee Im still seeing build errors - https://github.com/blocknative/web3-onboard/actions/runs/9503906496/job/26195385651?pr=2188

Adamj1232 reviewed (approved):
Approved! Looks good! We really appreciate your contribution to web3-onboard 🚀
